File: app/src/main/java/com/ianfield/bodyscoring/widgets/RecordAdapter.java
package com.ianfield.bodyscoring.widgets;

import android.support.annotation.Nullable;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.ianfield.bodyscoring.R;
import com.ianfield.bodyscoring.models.Record;

import java.text.SimpleDateFormat;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.RealmResults;

/**
 * Created by Ian Field on 1/20/16.
 */
public class RecordAdapter extends RecyclerView.Adapter<RecordAdapter.ViewHolder> {
    RealmResults<Record> records;
    OnRecordActionListener listener;
    SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");
    public RecordAdapter(@Nullable RealmResults<Record> records, @Nullable OnRecordActionListener listener) {
        this.records = records;
        this.listener = listener;
    }

    public RealmResults<Record> getRecords() {
        return records;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.record_view, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        final Record record = records.get(position);
        holder.name.setText(record.getName());
        holder.date.setText("Recorded: " + dateFormat.format(record.getScoringDate()));
        holder.plannedCalving.setText("Planned: " + dateFormat.format(record.getPlannedCalvingDate()));
        if (listener != null) {
            holder.view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onView(record.getId(), holder.name, holder.date, holder.plannedCalving);
                }
            });
            holder.edit.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onEdit(record.getId());
                }
            });
            holder.delete.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onDelete(record, holder.getAdapterPosition());
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return records == null ? 0 : records.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public CardView cardView;
        @BindView(R.id.name) public TextView name;
        @BindView(R.id.view) public Button view;
        @BindView(R.id.edit) public Button edit;
        @BindView(R.id.delete) public Button delete;
        @BindView(R.id.date) public TextView date;
        @BindView(R.id.planned_calving) public TextView plannedCalving;
        public ViewHolder(View container) {
            super(container);
            cardView = (CardView) container;
            ButterKnife.bind(this, container);
        }
    }

    public interface OnRecordActionListener {
        void onView(String id, TextView name, TextView date, TextView plannedCalving);

        void onEdit(String id);

        void onDelete(Record record, int position);
    }
}

File: app/src/main/java/com/ianfield/bodyscoring/widgets/RecordAdapter.kt
package com.ianfield.bodyscoring.widgets

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.ianfield.bodyscoring.R
import com.ianfield.bodyscoring.models.Record
import io.realm.RealmResults
import kotlinx.android.synthetic.main.record_view.view.*
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by Ian Field on 1/20/16.
 */
class RecordAdapter(val records: RealmResults<Record>?, private val listener: OnRecordActionListener?) : RecyclerView.Adapter<RecordAdapter.ViewHolder>() {
    private val dateFormat = SimpleDateFormat("dd/MM/yyyy", Locale.UK)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.record_view, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val record = records!![position]
        val context = holder.itemView.context
        holder.itemView.name.text = record?.name
        holder.itemView.date.text = context.getString(R.string.recorded_label,
                dateFormat.format(record?.scoringDate))
        holder.itemView.plannedCalving.text = context.getString(R.string.planned_calving_label,
                dateFormat.format(record?.plannedCalvingDate))
        if (listener != null) {
            holder.itemView.view.setOnClickListener { listener.onView(record?.id, holder.itemView.name, holder.itemView.date, holder.itemView.plannedCalving) }
            holder.itemView.edit.setOnClickListener { listener.onEdit(record?.id) }
            holder.itemView.delete.setOnClickListener { listener.onDelete(record, holder.adapterPosition) }
        }
    }

    override fun getItemCount(): Int {
        return records?.size ?: 0
    }

    interface OnRecordActionListener {
        fun onView(id: String?, name: TextView?, date: TextView?, plannedCalving: TextView?)

        fun onEdit(id: String?)

        fun onDelete(record: Record?, position: Int)
    }

    class ViewHolder(container: View) : RecyclerView.ViewHolder(container)
}
